File: archives/apple/shared/Core/API/APIClient.swift
import Foundation

/// Read-only client for activity listings (olympiads + textbooks).
///
/// The source of truth is archives/truth/olympiads.yml. A Python build
/// script generates archives/truth/olympiads.json which we fetch directly
/// from GitHub's raw content host. No database, no API layer, no writes.
public actor APIClient {
    public static let shared = APIClient()

    public static let baseURL = URL(
        string: "https://raw.githubusercontent.com/vivianweidai/science/main/archives/truth"
    )!

    private let session: URLSession
    private let decoder: JSONDecoder
    private var cachedActivities: [Activity]?
    private var cachedTopics: [ResearchTopic]?

    public init(session: URLSession = .shared) {
        self.session = session
        self.decoder = JSONDecoder()
    }

    /// Return cached activities if a previous fetch (e.g. the
    /// app-launch preload) already populated them; otherwise fetch.
    /// Caching here is what makes the background preload useful — when
    /// the user taps the Olympiads tab, its `.task` returns the
    /// already-loaded list instantly.
    public func listActivities() async throws -> [Activity] {
        if let cachedActivities { return cachedActivities }
        let items = try await get(file: "olympiads.json", as: ActivityList.self).items
        cachedActivities = items
        return items
    }

    public func listResearchTopics() async throws -> [ResearchTopic] {
        if let cachedTopics { return cachedTopics }
        let topics = try await get(file: "toys.json", as: ResearchToysResponse.self).topics
        cachedTopics = topics
        return topics
    }

    /// Invalidate caches — wired to pull-to-refresh so users can force
    /// a round trip when they've just pushed new YAML.
    public func invalidate() {
        cachedActivities = nil
        cachedTopics = nil
    }

    // MARK: - Private

    private func get<T: Decodable>(file: String, as: T.Type) async throws -> T {
        let url = Self.baseURL.appending(path: file)
        var request = URLRequest(url: url)
        request.cachePolicy = .reloadRevalidatingCacheData
        let (data, response) = try await session.data(for: request)
        try Self.check(response)
        return try decoder.decode(T.self, from: data)
    }

    private static func check(_ response: URLResponse) throws {
        guard let http = response as? HTTPURLResponse,
              (200..<300).contains(http.statusCode) else {
            throw URLError(.badServerResponse)
        }
    }
}

File: android/core/src/main/kotlin/com/vivianweidai/science/core/api/ApiClient.kt
package com.vivianweidai.science.core.api

import com.vivianweidai.science.core.model.Activity
import com.vivianweidai.science.core.model.ActivityList
import com.vivianweidai.science.core.model.ResearchTopic
import com.vivianweidai.science.core.model.ResearchToysResponse
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.serialization.json.Json

/**
 * Read-only client for activity listings (olympiads + textbooks) and
 * research toys. Source of truth: YAML in public/content/{olympiads,research}/,
 * built to JSON by a Python script, then fetched from vivianweidai.com.
 * No backend, no auth, no writes.
 */
class ApiClient {
    private val mutex = Mutex()
    private val json = Json { ignoreUnknownKeys = true }
    private var cachedActivities: List<Activity>? = null
    private var cachedTopics: List<ResearchTopic>? = null

    suspend fun listActivities(): List<Activity> = mutex.withLock {
        cachedActivities?.let { return it }
        val body = Http.getString(OLYMPIADS_URL)
        val list = json.decodeFromString<ActivityList>(body).items
        cachedActivities = list
        list
    }

    suspend fun listResearchTopics(): List<ResearchTopic> = mutex.withLock {
        cachedTopics?.let { return it }
        val body = Http.getString(TOYS_URL)
        val topics = json.decodeFromString<ResearchToysResponse>(body).topics
        cachedTopics = topics
        topics
    }

    suspend fun invalidate() = mutex.withLock {
        cachedActivities = null
        cachedTopics = null
    }

    companion object {
        const val OLYMPIADS_URL = "https://vivianweidai.com/content/olympiads/olympiads.json"
        const val TOYS_URL = "https://vivianweidai.com/content/research/toys.json"
        val shared = ApiClient()
    }
}
